### PR TITLE
CYS > Patterns Migration: Ensure a message is displayed if the user did not allow fetching patterns via API and/or they are unavailable

### LINKED ISSUE #47395: CYS > Patterns Migration: Ensure a message is displayed if the user did not allow fetching patterns via API and/or they are unavailable
**Issue Description:**

Ensure a message is displayed if:

* The site is not connected to the internet (local env)
* The PTK API is down
* The user didn't give us the required consent for allowing fetching the patterns via PTK

Designs:

FlPl7fHaGv9mpu8Nfl0joH-fi-3242_114763

### User did not give consent:

image

image

### User is offline

image

### PTK API is down

Copy TBD: @lauroraa could you drop a message here with the expected copy whenever you are available? Thanks!

### PR DESCRIPTION
**Pull request by albarin**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR shows a message on the assembler homepage section depending on the allow tracking option status.
It also adds a new private endpoint `wp-json/wc/private/patterns` that is used from the assembler state machine to trigger a fetch of patterns from the PTK API when the user opts-in on tracking.

Closes #47395 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to `wp-admin/admin.php?page=wc-settings&tab=advanced&section=woocommerce_com` and make sure `Allow usage of WooCommerce to be tracked` is **unchecked**.
2. Make sure you have the WooCommerce beta tester plugin installed.
3. Head over to `Tools > WCA Test Helper > Features`.
4. Make sure you see the `pattern-toolkit-full-composability` on the list and enable it.
5. Head over to `WooCommerce > Home > Customize your store`.
6. Click on `Design your homepage`.
7. Scroll to the bottom of the sidebar and make sure you see:
Screenshot 2024-06-04 at 11 45 53

8. Open the Chrome developer tools and on `Network` select `Offline`.
9. Make sure you now see:
Screenshot 2024-06-04 at 11 49 23

10. On the list of request search for `wp-json/wc/private/patterns?_locale=user`, right-click on it, and select `Block request URL`.
11. Refresh the page and make sure you now see:
Screenshot 2024-06-04 at 11 52 18

12. Unblock the request and refresh the page.
13. Scroll down and click on `usage tracking`, then `Opt-in`.
14. Wait for the process to return to the assembler, and click on `Design your homepage`.
15. Scroll to the bottom and make sure you don't see any of the messages above anymore.
16. Exit the assembler and create a new post or page.
17. Make sure you can insert the `Services: Three column pricing table` pattern..

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
[CYS] Show a message when tracking is not allowed in patterns and add the ability for users to opt-in and fetch patterns.

</details>

<details>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-homepage.tsx [hunk 1]
```diff
@@ -223,6 +228,51 @@ export const SidebarNavigationScreenHomepage = () => {
...
+			'woocommerce'
+		);
+	}
+
+	const [ isModalOpen, setIsModalOpen ] = useState( false );
+
+	const openModal = () => setIsModalOpen( true );
+	const closeModal = () => setIsModalOpen( false );
+
+	const [ OptInDataSharing, setIsOptInDataSharing ] =
```
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-homepage.tsx]: Small typo
```suggestion
	const [ optInDataSharing, setIsOptInDataSharing ] =
```

**Review by nefeline — APPROVED** [review]
Thanks for working on this @albarin ! I can confirm the implementation works as described for all use-cases, including:

✅ dedicated message for when the user didn't allow usage tracking:
Screenshot 2024-06-05 at 12 34 36

✅ Dedicated message for when the user is offline:

Screenshot 2024-06-05 at 12 35 25

✅ Dedicated message for when the PTK API is unavailable:
Screenshot 2024-06-05 at 12 37 16

✅ I've also confirmed the modal to allow the usage tracking is in place: clicking on opt-in refreshes the page and lead the user back to the assembler and in this case, no messages are displayed.

I just have one small suggestion that could be worth exploring: how about we ensure the users land back to where they were, back to "Design your homepage" after they clicked to allow usage tracking and the assembler is refreshed?

**Comment by albarin** [comment]
> I just have one small suggestion that could be worth exploring: how about we ensure the users land back to where they were, back to "Design your homepage" after they clicked to allow usage tracking and the assembler is refreshed?

Thanks for you review @nefeline Yes, I think that makes sense, wdyt about merging this and opening a new issue to explore this for patterns and also for fonts? Both follow similar flows and I think it would make sense to implement it on both. I can create the issue.

**Comment by nefeline** [comment]
> wdyt about merging this and opening a new issue to explore this for patterns and also for fonts? Both follow similar flows and I think it would make sense to implement it on both. I can create the issue.

Sounds good to me, thanks! 👍

**Review by gigitux — APPROVED** [review]
Thanks for working on this! 🚀 .  Everything LGTM. I left a comment about a typo.

As we discussed on Slack (p1717513346982719-slack-C053716F2H2), we came up with that the action scheduler can't guarantee that that action for setting the transient will be triggered in time. For this reason, we should add another check during the setup of the store to check if the transient is set, if not, we have to set them via `install_patterns function`. We can create a dedicated issue/PR for this task! 

Thanks again!

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-homepage.tsx (hunk 1 above)
Comment (albarin) [reply plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-homepage.tsx]: Fixed here https://github.com/woocommerce/woocommerce/pull/48095/commits/722dcac7bdd843af8877ab946ef8abcde06df3b7

**Comment by nefeline** [comment]
> we came up with that the action scheduler can't guarantee that that action for setting the transient will be triggered in time. For this reason, we should add another check during the setup of the store to check if the transient is set, if not, we have to set them via install_patterns function. We can create a dedicated issue/PR for this task!

The transient is set [on init](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Blocks/BlockPatterns.php#L87) and it is triggered basically on every single page load within WordPress, so it is very unlikely it won’t be finished by the time the user reaches the assembler, but it doesn’t hurt adding this extra check if you want to, although I suspect it will be rarely triggered at this point.

**Pull request merged**